Enhancing the guard command. Guarding units will form a circle around protected unit. Will take heading of protected unit and enemies into account

roberthartmann opened the pull request:
### Including an adaptet version of circle guard

stealing from ZK

current goals:
1) create a circle around the protected unit
2) begin circle in front of the unit (look at unit direction)
3)  if enemies are close or protected unit takes damage begin to fill up the circle in that direction 

How the circle is meant to be filled
1) first unit in front
2) 2 units 120° spread. Each one 60° to the front left and right of the protected unit
3) 3+ units 360°/n guards = separation between guards
 a) uneven numbers: one unit at the front
 b) even numbers 2 units in front (equal spread left tight of front)

roberthartmann committed:
Create cmd_areaguard.lua

roberthartmann committed:
Create cmd_orbit.lua

roberthartmann commented:
when I try to add the vector utilities everything stops working

https://github.com/ZeroK-RTS/Zero-K/blob/9fbbbb9e469586d205a60867b691fd8ca6380ae3/LuaRules/Utilities/vector.lua#L261

ZephyrSkies7 commented:
Please rename the PR to something more professional and reflective of the actual changes

sprunk commented on luarules/gadgets/cmd_areaguard.lua:
```diff
@@ -0,0 +1,430 @@
+
+local cos = math.cos
+local sin = math.sin
+local pi = math.pi
```
almost always used with a manual x2 multiplication below, consider tau instead

sprunk commented on luarules/gadgets/cmd_areaguard.lua:
```diff
@@ -0,0 +1,430 @@
+--------------------------------------------------------------------------------
+-- Functions
+
+local function RandomPermutation(n)
```
sounds like an excellent thing to put into table utilities, if not there already

sprunk commented on luarules/gadgets/cmd_areaguard.lua:
```diff
@@ -0,0 +1,430 @@
+		-- Factor of 3 acts as leeway because if a unit gets stuck nothing is
+		-- going to move it.
+		angle = myAngle + 3*circleDirection[targetID]*CIRC_MULT/circuitTime
+		perpSize = circleDirection[targetID]*50
```
what's the 50? could use a well-named constant

roberthartmann replied:
magic numbers everywhere. I know. It's the original

roberthartmann committed:
updated version still doesn't work

I know it's not good, but I don't want you to spend time on the even older version

sprunk commented on luaui/Widgets/cmd_orbit.lua:
```diff
@@ -0,0 +1,73 @@
+
+--widget:UnitCommand(unitID, unitDefID, unitTeam, cmdID, cmdParams, cmdOpts, cmdTag)
+function widget:CommandNotify(cmdID, params, options)
+	if (cmdID == 25) and (#params == 1) and options.ctrl then
```
use CMD.BLA for the 25

sprunk commented on luarules/gadgets/cmd_areaguard.lua:
```diff
@@ -0,0 +1,430 @@
+	return {[CMD_GUARD] = true, [CMD_AREA_GUARD] = true, [CMD_ORBIT] = true, [CMD_ORBIT_DRAW] = true}
+end
+
+function gadget:AllowCommand_GetWantedUnitDefID()
```
BAR uses some different setup than ZK for allowcommand

roberthartmann commented:
will look at all of your comments.
need to understand the widget and gadgets better.

roberthartmann committed:
Delete luarules/gadgets/unit_fast_turrets.lua

it's not part of this PR

roberthartmann committed:
Update modoptions.lua

fast turrets are not part of this PR

roberthartmann commented on luarules/gadgets/cmd_areaguard.lua:
```diff
@@ -0,0 +1,430 @@
+--local SUC = spUtilities.CMD
+
+local env = getfenv()
+local SUC = VFS.Include("LuaRules/Configs/customcmds.lua", nil, VFS.GAME)
+for cmdName, cmdID in pairs(SUC) do
+	env["CMD_" .. cmdName] = cmdID
+end
+env.CMD_SETHAVEN = SUC.RETREAT_ZONE
```
ah, I see.
That is why I needed this workaround in gadget.lua, I guess.

thx!

roberthartmann commented on luarules/gadgets/cmd_areaguard.lua:
```diff
@@ -0,0 +1,430 @@
+
+function gadget:AllowCommand(unitID, unitDefID, unitTeam, cmdID, cmdParams, cmdOptions)
+	if cmdID == CMD_GUARD then
+		if #cmdParams == 1 and unitID == cmdParams[1] then
```
you might need the ```#params ==1``` to make sure that you don't read out part of the coordinates.

roberthartmann commented:
miss clicked
